grafana-plugin/integration-tests/globalSetup.ts

```diff
@@ -1,37 +1,57 @@
-import { test as setup, chromium, FullConfig, expect, Page, BrowserContext, APIResponse } from '@playwright/test';
-
-import { BASE_URL, GRAFANA_PASSWORD, GRAFANA_USERNAME, IS_OPEN_SOURCE, ONCALL_API_URL } from './utils/constants';
+import { test as setup, chromium, expect, Page, BrowserContext, FullConfig, APIRequestContext } from '@playwright/test';
+
+import GrafanaAPIClient from './utils/clients/grafana';
+import {
+  GRAFANA_ADMIN_PASSWORD,
+  GRAFANA_ADMIN_USERNAME,
+  GRAFANA_EDITOR_PASSWORD,
+  GRAFANA_EDITOR_USERNAME,
+  GRAFANA_VIEWER_PASSWORD,
+  GRAFANA_VIEWER_USERNAME,
+  IS_CLOUD,
+  IS_OPEN_SOURCE,
+  ONCALL_API_URL,
+} from './utils/constants';
 import { clickButton, getInputByName } from './utils/forms';
 import { goToGrafanaPage } from './utils/navigation';
-import { STORAGE_STATE } from '../playwright.config';
+import { VIEWER_USER_STORAGE_STATE, EDITOR_USER_STORAGE_STATE, ADMIN_USER_STORAGE_STATE } from '../playwright.config';
+import { OrgRole } from '@grafana/data';
 
-const IS_CLOUD = !IS_OPEN_SOURCE;
-const GLOBAL_SETUP_RETRIES = 3;
+const grafanaApiClient = new GrafanaAPIClient(GRAFANA_ADMIN_USERNAME, GRAFANA_ADMIN_PASSWORD);
 
-const makeGrafanaLoginRequest = async (browserContext: BrowserContext): Promise<APIResponse> =>
-  browserContext.request.post(`${BASE_URL}/login`, {
-    data: {
-      user: GRAFANA_USERNAME,
-      password: GRAFANA_PASSWORD,
-    },
-  });
+type UserCreationSettings = {
+  adminAuthedRequest: APIRequestContext;
+  role: OrgRole;
+};
 
-const pollGrafanaInstanceUntilItIsHealthy = async (browserContext: BrowserContext): Promise<boolean> => {
-  console.log('Polling the grafana instance to make sure it is healthy');
+const generateLoginStorageStateAndOptionallCreateUser = async (
+  config: FullConfig,
+  userName: string,
+  password: string,
+  storageStateFileLocation: string,
+  userCreationSettings?: UserCreationSettings,
+  closeContext = false
+): Promise<BrowserContext> => {
+  if (userCreationSettings !== undefined && IS_OPEN_SOURCE) {
+    const { adminAuthedRequest, role } = userCreationSettings;
+    await grafanaApiClient.idempotentlyCreateUserWithRole(adminAuthedRequest, userName, password, role);
```

Comment:
Hmm shouldn't this go in a try catch clause?

Reply:
I think it's okay here. The `playwright.APIRequestContext` object that makes the HTTP requests operates a bit different from `axios` that we use for the actual plugin UI. It doesn't throw exceptions for HTTP 412 (user already exists), whereas `axios` probably does?

Reply:
Hmm that's odd, as you're also throwing an exception from that function if you get some other response back.

Reply:
ah fair point 👍 in that case I think the tests should purposefully fail, because something is not setup the way we would expect